### ISSUE TITLE
mountebank extensions

### ISSUE THREAD

**Issue by moodmosaic** [bbyars/mountebank#14]
tl;dr
_A hypothetical extension for mountebank could supply over the wire Equivalence Sets for primary data types (String, Number, Boolean) and composite data types (Object, Array)._
#### Motivation

Test Doubles (Mocks, Spies, Stubs, etc) often play an important role in test robustness – another concept in test robustness is the [Equivalence Classes](http://xunitpatterns.com/equivalence%20class.html) (or Equivalence Sets).
#### Pareto principle

Imagine, for example, a [SUT](http://xunitpatterns.com/SUT.html) that accepts a direct input value and _if_ the input value is > 1 and < 100 the _expected_ value is always equal to _actual_ result value.

In this example, _any_ number from the Equivalence Set [1 .. 100] is guaranteed to exercise the same path of the code thus making the test more _robust_.

In [AutoFixture](https://github.com/AutoFixture/AutoFixture) we have generalized this approach by the _[80–20 rule](http://en.wikipedia.org/wiki/Pareto_principle)_:

_In most mainstream scenarios, numbers of the Equivalence Set [1 .. 255] tend to exercise the same path in the code._
#### In practise

The same way we create an imposter – we could ask a mountebank **extension** to create a [fixture](http://xunitpatterns.com/test%20fixture%20-%20xUnit.html) and transmit values for requested data types drained from an Equivalence Set over the wire.

A good collection of Equivalence Sets for primary data types (String, Number, Boolean) gives the ability to create _composite_ data types (Object, Array).
#### Example

One could request the following JSON Schema from a mountebank extension:

```
{
    "name": "Product",
    "properties": {
        "id": {
            "type": "number",
            "description": "Product identifier",
            "required": true
        },
        "name": {
            "type": "string",
            "description": "Name of the product",
            "required": true
        },
        "price": {
            "type": "number",
            "minimum": 0,
            "required": true
        },
       "datetime": {
            "type": "date"
        },
        "tags": {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        "stock": {
            "type": "object",
            "properties": {
                "warehouse": {
                    "type": "number"
                },
                "retail": {
                    "type": "number"
                }
            }
        }
    }
}
```

And the mountebank extension would respond with a JSON generalized by the 80–20 rule:

```
{
    "id": 50,
    "name": "89071263-eeb2-4a1d-84bf-31c760360bc9",
    "price": 123,
    "datetime": "2014-05-10T00:18:40.055Z" 
    "tags": [ "Bar-89071263-eeb2", "Eek-89071263-eeb2" ],
    "stock": {
        "warehouse": 300,
        "retail": 20
    }
}
```

> It should be also possible to create custom Equivalence Sets (via HTTP) by specifying the [Boundary Values](http://xunitpatterns.com/boundary%20value.html) in the JSON string.
#### Benefits

Mountebank could eventually become the _de facto_, truly cross-platform, tool around test robustness, capable of creating Test Doubles, and also Fixtures via extensions using Equivalence Sets generalized by the 80–20 rule.
#### References
- [xUnit Test Patterns](http://xunitpatterns.com/)
- [AutoFixture](https://github.com/AutoFixture)
- [Truly constrained non-deterministic numbers](https://github.com/AutoFixture/AutoFixture/issues/2)
- [Herding Code 165: Mark Seemann on AutoFixture and Unit Testing](http://herdingcode.com/herding-code-165-mark-seemann-on-autofixture-and-unit-testing/)

---

If this hypothetical mountebank extension falls into a direction you would like to take, please let me know. I may work on it, in my spare time, and I may open a pull request sometime.
